extensions/aws/processors/FetchS3Object.cpp

```diff
@@ -0,0 +1,122 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "FetchS3Object.h"
+
+#include <set>
+#include <memory>
+
+namespace org {
+namespace apache {
+namespace nifi {
+namespace minifi {
+namespace aws {
+namespace processors {
+
+const core::Property FetchS3Object::Version(
+  core::PropertyBuilder::createProperty("Version")
+    ->withDescription("The Version of the Object to download")
+    ->supportsExpressionLanguage(true)
+    ->build());
+const core::Property FetchS3Object::RequesterPays(
+  core::PropertyBuilder::createProperty("Requester Pays")
+    ->isRequired(true)
+    ->withDefaultValue<bool>(false)
+    ->withDescription("If true, indicates that the requester consents to pay any charges associated with retrieving "
+                      "objects from the S3 bucket. This sets the 'x-amz-request-payer' header to 'requester'.")
+    ->build());
+
+const core::Relationship FetchS3Object::Success("success", "FlowFiles are routed to success relationship");
+const core::Relationship FetchS3Object::Failure("failure", "FlowFiles are routed to failure relationship");
+
+void FetchS3Object::initialize() {
+  // Add new supported properties
+  updateSupportedProperties({Version, RequesterPays});
+  // Set the supported relationships
+  setSupportedRelationships({Failure, Success});
+}
+
+void FetchS3Object::onSchedule(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSessionFactory> &sessionFactory) {
+  S3Processor::onSchedule(context, sessionFactory);
+
+  context->getProperty(RequesterPays.getName(), get_object_params_.requester_pays);
+  logger_->log_debug("FetchS3Object: RequesterPays [%s]", get_object_params_.requester_pays ? "true" : "false");
+}
+
+bool FetchS3Object::getExpressionLanguageSupportedProperties(
+    const std::shared_ptr<core::ProcessContext> &context,
+    const std::shared_ptr<core::FlowFile> &flow_file) {
+  if (!S3Processor::getExpressionLanguageSupportedProperties(context, flow_file)) {
+    return false;
+  }
+  get_object_params_.bucket = std::move(bucket_);
+  get_object_params_.object_key = std::move(object_key_);
+
+  context->getProperty(Version, get_object_params_.version, flow_file);
+  logger_->log_debug("FetchS3Object: Version [%s]", get_object_params_.version);
+  return true;
+}
+
+void FetchS3Object::onTrigger(const std::shared_ptr<core::ProcessContext> &context, const std::shared_ptr<core::ProcessSession> &session) {
+  logger_->log_debug("FetchS3Object onTrigger");
+  std::shared_ptr<core::FlowFile> flow_file = session->get();
+  if (!flow_file) {
+    context->yield();
+    return;
+  }
+
+  if (!getExpressionLanguageSupportedProperties(context, flow_file)) {
+    session->transfer(flow_file, Failure);
+    return;
+  }
+
+  WriteCallback callback(flow_file->getSize(), get_object_params_, s3_wrapper_.get());
+  session->write(flow_file, &callback);
```

Comment:
I wonder if it may throw anyhow. Or just fails?
I don't say that the code is wrong, just would like to double-check to make sure we don't execute rollbacks in case there is an issue with one specific flowfile.

Reply:
After checking the code there seems to be no way for this to throw. The callback can only fail and return 0 if failure happens. The write may throw an exception only if the return value of the callback is less than 0.

Reply:
Okay, thanks.